File: Solutions/Python/ipo.py
"""
    [HARD]
    502. IPO

    Concepts:
    - array
    - greedy
    - heap/priority queue

    Stats:
        Runtime | 377 ms    [Beats 44.77%]
        Memory  | 42.10 MB  [Beats 82.38%]
"""

class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        money = w
        max_prof = []
        min_cap = list(zip(capital, profits))
        heapq.heapify(min_cap)

        for _ in range(k):
            # push all affordable projects to max heap `max_prof`
            while min_cap and min_cap[0][0] <= money:
                c, p = heapq.heappop(min_cap)
                heapq.heappush(max_prof, -p)

            # if no affordable projects...
            if not max_prof:
                break

            # complete most profitable project
            money -= heapq.heappop(max_prof)

        return money

File: Solutions/Cpp/ipo.cpp
/*
    [HARD]
    502. IPO

    Concepts:
    - array
    - greedy
    - heap/priority queue

    Stats:
        Runtime | 104 ms    [Beats 28.63%]
        Memory  | 131.30 MB [Beats 74.32%]
*/

class Solution {
public:
    int findMaximizedCapital(int k, int money, vector<int>& profits, vector<int>& capital) {
        std::priority_queue<int> max_prof = {};
        std::priority_queue<
            std::pair<int, int>,
            std::vector<std::pair<int, int>>,
            std::greater<std::pair<int, int>>
        > min_cap = {};

        for (int i = 0; i < profits.size(); i++)
            min_cap.emplace(capital[i], profits[i]);

        for (int _ = 0; _ < k; _++) {
            // push all affordable projects to max heap 'min_prof'
            while (!min_cap.empty() && min_cap.top().first <= money) {
                max_prof.push(min_cap.top().second);
                min_cap.pop();
            }

            // if no affordable projects...
            if (max_prof.empty())
                break;

            // complete most profitable project
            money += max_prof.top(); max_prof.pop();
        }

        return money;
    }
};